### ISSUE TITLE
Missing KL diverge term in the loss for the VAE?

### ISSUE THREAD

**Issue by karpathy** [lucidrains/DALLE-pytorch#40]
Thanks for the repo! Was skimming around and found

https://github.com/lucidrains/DALLE-pytorch/blob/c27f48c4373942167a0cc2ac2238f67af2479fe3/dalle_pytorch/dalle_pytorch.py#L158

where I'm a bit nervous that you may be missing the KL Divergence term to the prior in the -ELBO, which in this case would be a diffuse uniform.

- Eric's tutorial [here](https://blog.evjang.com/2016/11/tutorial-categorical-variational.html)
- Eric's implementation [here ](https://github.com/ericjang/gumbel-softmax/blob/master/gumbel_softmax_vae_v2.ipynb) implements this under `KL_qp`
- Another random implementation [here](https://github.com/dev4488/VAE_gumble_softmax/blob/master/vae_gumbel_softmax.py#L129)
- [My own](https://github.com/karpathy/deep-vector-quantization/blob/main/model.py#L98) implementation when I was deriving this today. Though it skips a few steps of derivation, could have been clearer 

Without this term you're potentially going to see "index collapse", with many of the dimensions going unused.

In any case, in my preliminary experiments I'm finding this version a little bit finicky to tune so far, the VQVAE version worked more more out of the box and trained faster for me, though only after a bit gross data-dependent init scheme.

**Comment by lucidrains** [comment]
@karpathy Hi Andrej! Big fan (I'm sure you get that a lot) :)

Thank you so much for the code review; I've put out a patch https://github.com/lucidrains/DALLE-pytorch/commit/29e40a399bc47fdd938e8649e497fb0d43bbddac I'll actually be starting to train DALL-E this week as well, so your advice comes at a good time!

**Comment by karpathy** [comment]
Ok great! Glad to contribute, impressed with the speed with which you're implementing and releasing all of these repos!

**Comment by CDitzel** [comment]
can we talk again please about the KL term? Leaving it out gives far better results than using it with whatever schedule or parameter range imaginable. The authors were quite evasive in the clubhouse discussion with Andrej and Justin. The paper mentions an increasing schedule but I could not get it to work, even after two weeks of trying. 

Are we sure that the prior is uniform? With decreasing temperature of the Gumbel Softmax, the distribution parameterized by the encoder becomes increasingly non uniform doesnt it? so why would we encourage it to be uniform in the first place?

Im at a complete loss here
